CI: shift core-library tests to a separate "job"

mabruzzo opened the pull request:
This PR shifts the core-library tests to a separate "job."

It has come to my attention in recent PRs (e.g. #273, #279), that our current choice to run the core-library test-suite after the Pygrackle test-suite is somewhat undesirable. Specifically, the core-library test-suite is more likely to include unit-tests (while there may not be many unit-tests yet, many forthcoming PRs have proposed adding them).

This is important for 2 reasons:
1. Unit-tests are inherently faster than answer-tests. We should run these as soon as possible so we can fail quickly.
2. Unit-tests are also more specialized than answer-tests. If a unit-test fails, it is likely that an answer-test will also fail. The way things have been configured (before this PR) will cause the answer-test to fail and CI to abort before the unit-tests can be reun.

As I write this up, I realize that I probably just could have re-ordered the tests. I'm totally willing to do that. But, I think in some ways this may be better (if we aren't too worried about using up resources)

brittonsmith reviewed (approved):
This looks good. I had one minor comment. Feel free to deal with that and merge. Only slightly related to this, one thing I don't like is that we run tests with the classic build system in generate mode prior to creating the gold standard. It probably had marginal benefit, but I think it would make sense for gold standard generation to happen first and for as much of the testing as is possible to run in compare mode.

mabruzzo commented on tests/scripts/code_example_checker.py:
```diff
@@ -264,40 +265,132 @@ def main_cmp(args):
     atol, rtol = args.atol, args.rtol
     print(f"Comparing Results: atol = {atol}, rtol = {rtol}")
```
Thanks for catching this

mabruzzo commented:
> Only slightly related to this, one thing I don't like is that we run tests with the classic build system in generate mode prior to creating the gold standard. It probably had marginal benefit, but I think it would make sense for gold standard generation to happen first and for as much of the testing as is possible to run in compare mode.

I 100% agree. I've been tempted to change this a couple of times. But, I was previously under the false impression that you had picked the current order (I now suspect that the current order originated from a typo OR the fact that you and I were making large changes to the answer testing framework around the same time as each other)

mabruzzo commented:
As we discussed offline, I adjusted the continuous integration so that now run tests with the classic build system in *compare-mode* **AFTER** we record the test-answers.

While I doing this I made 3 really minor tweaks:
- I fixed a typo in the name of one of the new CI "steps"
- I made sure the CI will explicitly fail if we make a particular mistake when modify `config.yml` (we were already printing an error message for this case)

More importantly, I caught a bug where we weren't properly running the gold-standard in compare mode (this was due to a typo on my part in a previous PR). While trying to fix that bug, I found and fixed a separate minor bug that was messing with string comparisons when we use circle-ci's parameters (as luck would have it, I actually think everything was previously working as intended)
